### PR TITLE
Support Writing CRD Status Subresource in K8S Client

### PR DESCRIPTION
**Pull request by adiprerepa**
Add status field as part of the vm healthchecking effort.

cc @howardjohn

### CODE REVIEW

File: pilot/pkg/config/kube/crdclient/types.gen.go [hunk 1]
```diff
@@ -47,56 +48,67 @@ func create(ic versionedclient.Interface, sc serviceapisclient.Interface, cfg co
 		return ic.NetworkingV1alpha3().DestinationRules(cfg.Namespace).Create(context.TODO(), &clientnetworkingv1alpha3.DestinationRule{
 			ObjectMeta: objMeta,
 			Spec:       *(cfg.Spec.(*networkingv1alpha3.DestinationRule)),
+			Status:     *(cfg.Status.(*v1alpha1.IstioStatus)),
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/types.gen.go:51]: I think spec will be empty for most use cases? so are we going to get nil pointer dereferences here?

**Review by howardjohn — COMMENTED** [review]
Can you add some tests (we have some existing ones in the crd/ package) of reading and writing status?

**Comment by howardjohn** [comment]
cc @therealmitchconnors @jasonwzm

File: pilot/pkg/config/kube/crd/config.go [hunk 2]
```diff
@@ -24,13 +24,19 @@ type IstioKind struct {
 	meta_v1.TypeMeta   `json:",inline"`
 	meta_v1.ObjectMeta `json:"metadata"`
 	Spec               map[string]interface{} `json:"spec"`
+	Status             map[string]interface{} `json:"status"`
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crd/config.go:27]: maybe we should add `json:"status,omitempty"` to avoid the `status: null`?

File: pilot/pkg/config/kube/crd/conversion.go [hunk 3]
```diff
@@ -78,7 +78,15 @@ func ConvertObject(schema collection.Schema, object IstioObject, domain string)
...
-	data, err := FromJSON(schema, string(js))
+	spec, err := FromJSON(schema, string(js))
+	if err != nil {
+		return nil, err
+	}
+	js, err = json.Marshal(object.GetStatus())
+	if err != nil {
+		return nil, err
+	}
+	status, err := FromJSON(schema, string(js))
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crd/conversion.go:89]: Can we make sure status ends up as `nil` not an empty (but initialized) map when status is not set? To avoid extra allocations

File: pilot/pkg/config/kube/crd/conversion_test.go [hunk 4]
```diff
@@ -38,9 +39,11 @@ func TestConvert(t *testing.T) {
 			Annotations:      map[string]string{"annotation": "value"},
 		},
 		Spec: mock.ExampleVirtualService,
+		Status: map[string]interface{}{},
 	}
 
 	obj, err := ConvertConfig(config)
+	log.Println(obj.GetStatus())
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crd/conversion_test.go:46]: we can remove this before merge

File: pilot/pkg/config/kube/crdclient/client.go [hunk 5]
```diff
@@ -205,12 +205,18 @@ func (cl *Client) Get(typ config.GroupVersionKind, name, namespace string) *conf
...
-		return "", fmt.Errorf("nil spec for %v/%v", config.Name, config.Namespace)
+func (cl *Client) Create(cfg config.Config) (string, error) {
+	if cfg.Spec == nil {
+		return "", fmt.Errorf("nil spec for %v/%v", cfg.Name, cfg.Namespace)
+	}
+
+	// we probably dont want to error here because its nil in most
+	// use cases. just populate it with an empty status.
+	if cfg.Status == nil {
+		cfg.Status = new(config.Status)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client.go:216]: why can't we keep it nil?

File: pilot/pkg/config/kube/crdclient/types.gen.go (hunk 1 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crdclient/types.gen.go:51]: do you mean status?

spec was there before

File: pilot/pkg/config/kube/crdclient/client.go (hunk 5 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crdclient/client.go:216]: i thought when create() would later be called and a nil pointer dereference would happen:

```
			Status:     *(cfg.Status.(*v1alpha1.IstioStatus))
```

File: pilot/pkg/config/kube/crd/config.go [hunk 6]
```diff
@@ -42,6 +48,7 @@ func (in *IstioKind) DeepCopyInto(out *IstioKind) {
 	out.TypeMeta = in.TypeMeta
 	in.ObjectMeta.DeepCopyInto(&out.ObjectMeta)
 	out.Spec = in.Spec
+	out.Status = in.Status
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crd/config.go:51]: Just find that since the beginning, this function is wrong. It does not really deepcopy.

File: pilot/pkg/config/kube/crdclient/types.gen.go (hunk 1 above)
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/types.gen.go:51]: I kind of remember `status` will be ignored when create, can you confirm with it.

Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/types.gen.go:51]: Found one specification: https://book-v1.book.kubebuilder.io/basics/status_subresource.html

Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/types.gen.go:51]: yes sorry

Comment (adiprerepa) [reply pilot/pkg/config/kube/crdclient/types.gen.go:51]: yeah great point. only on update() it can be changed.

File: pilot/pkg/config/kube/crd/conversion.go (hunk 3 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crd/conversion.go:89]: i think since we know its an `IstioCondition` we are marshalling into we dont need that function altogether

File: pilot/pkg/config/kube/gateway/testdata/simple.yaml [hunk 7]
```diff
@@ -109,3 +113,4 @@ spec:
       forwardTo:
       - targetRef:
           name: httpbin
+status: null
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/gateway/testdata/simple.yaml:116]: can we remove

File: pilot/pkg/config/kube/crdclient/client_test.go [hunk 8]
```diff
@@ -129,8 +131,9 @@ func TestClient(t *testing.T) {
 			}
 
 			if _, err := store.Create(config.Config{
-				Meta: configMeta,
-				Spec: pb,
+				Meta:   configMeta,
+				Spec:   pb,
+				Status: &v1alpha1.IstioStatus{},
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client_test.go:136]: remove

File: pilot/pkg/config/kube/crdclient/client_test.go [hunk 9]
```diff
@@ -79,8 +80,9 @@ func TestClientNoCRDs(t *testing.T) {
 	}
 
 	if _, err := store.Create(config.Config{
-		Meta: configMeta,
-		Spec: pb,
+		Meta:   configMeta,
+		Spec:   pb,
+		Status: &v1alpha1.IstioStatus{},
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crdclient/client_test.go:85]: same here

**Comment by hzxuzhonghu** [comment]
A release note is necessary

File: pilot/pkg/config/kube/crd/conversion_test.go [hunk 10]
```diff
@@ -38,11 +39,17 @@ func TestConvert(t *testing.T) {
...
+				{Type: "Health"},
+			},
+		},
 	}
 
 	obj, err := ConvertConfig(config)
 	if err != nil {
-		t.Errorf("ConvertConfig() => unexpected error %v", err)
+		t.Errorf("Conv"+
+			"ertConfig() => unexpected error %v", err)
```
Comment (shamsher31) [inline pilot/pkg/config/kube/crd/conversion_test.go:52]: Any reason not to directly write `ConvertConfig()`?

File: pilot/pkg/config/kube/crd/conversion.go [hunk 11]
```diff
@@ -45,6 +46,22 @@ func FromJSON(s collection.Schema, js string) (config.Spec, error) {
 	return c, nil
 }
 
+func IstioStatusJSONFromMap(jsonMap map[string]interface{}) (config.Status, error) {
```
Comment (shamsher31) [inline pilot/pkg/config/kube/crd/conversion.go:49]: use `StatusJSONFromMap` instead of `IstioStatusJSONFromMap`?

File: pilot/pkg/config/kube/gateway/conversion_test.go [hunk 12]
```diff
@@ -46,6 +47,7 @@ func TestConvertResources(t *testing.T) {
 			}
 			golden := splitOutput(readConfig(t, goldenFile))
 			if diff := cmp.Diff(golden, output); diff != "" {
+				log.Println(output)
```
Comment (shamsher31) [inline pilot/pkg/config/kube/gateway/conversion_test.go:50]: Please remove this if it is no longer needed.

File: pilot/pkg/config/kube/crd/conversion_test.go (hunk 10 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crd/conversion_test.go:52]: that was a mistake lol

File: pilot/pkg/config/kube/crd/conversion.go (hunk 11 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crd/conversion.go:49]: Its pretty specific to `IstioStatus`, and there are other types of statuses.

File: pilot/pkg/config/kube/crd/config_test.go [hunk 13]
```diff
@@ -32,6 +32,12 @@ func TestKind(t *testing.T) {
 		t.Errorf("GetSpec() => got %v, want %v", got, spec)
 	}
 
+	status := map[string]interface{}{"yo": "lit", "istio>linkerd": "true"}
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crd/config_test.go:35]: Please don't put comments like `istio>linkerd`

File: pilot/pkg/config/kube/crd/conversion.go [hunk 14]
```diff
@@ -78,10 +95,14 @@ func ConvertObject(schema collection.Schema, object IstioObject, domain string)
...
 		return nil, err
 	}
-	data, err := FromJSON(schema, string(js))
+	spec, err := FromJSON(schema, string(js))
 	if err != nil {
 		return nil, err
 	}
+	status, err := IstioStatusJSONFromMap(object.GetStatus())
+	if err != nil {
+		log.Errorf("could not get istio status from map %v, err %v", object.GetStatus(), err)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crd/conversion.go:104]: Any reason to log the error here but the previous one we will return an error?

File: pilot/pkg/config/kube/crdclient/gen/main.go [hunk 15]
```diff
@@ -51,6 +53,7 @@ type ConfigData struct {
 func MakeConfigData(schema collection.Schema) ConfigData {
 	out := ConfigData{
 		Namespaced:      !schema.Resource().IsClusterScoped(),
+		IstioStatus:     !strings.Contains(schema.VariableName(), "K8S"),
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/gen/main.go:56]: There are k8s types we will want to handle status for as well. Do we need to filter? I think everything is standardizing on https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#Condition across the ecosystem anyways?

@therealmitchconnors

**Review by howardjohn — COMMENTED** [review]
Can you add test actually creating resources with status? I don't think it will pass as is

File: pilot/pkg/config/kube/crd/conversion.go (hunk 14 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crd/conversion.go:104]: i dont think we want status breaking/erroring if something goes wrong, right? could be wrong

File: pilot/pkg/config/kube/crdclient/gen/main.go (hunk 15 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/crdclient/gen/main.go:56]: i mean if i dont filter an error happens because the k8s resources statuses are not `IstioStatus`.

for example, `TCPRoute` has `TCPRouteStatus`, not `IstioStatus`.

File: pilot/pkg/config/kube/crdclient/client_test.go [hunk 16]
```diff
@@ -173,4 +173,41 @@ func TestClient(t *testing.T) {
...
+			Meta:   wgConfigMeta,
+			Spec:   spec,
+			Status: stat,
+		})
+		if err != nil {
+			t.Errorf("err: %v", err)
+			return err
+		}
+		return nil
+	}, timeout)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client_test.go:212]: Can we do a GET and verify status is actually set? If I member correctly, we need to explicitly call UpdateStatus not just Update for status to be changed

File: pilot/pkg/config/kube/crdclient/gen/main.go (hunk 15 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/gen/main.go:56]: How does this work now that only WorkloadEntry has status defined?

Comment (adiprerepa) [reply pilot/pkg/config/kube/crdclient/gen/main.go:56]: I think you mean `WorkloadGroup`.

And that one change was just for testing stuff, I should be adding the status field to all of the Pilot Fields Now.

**Review by esnible — COMMENTED** [review]
I like the direction.  Please address howard's requests.

File: pilot/pkg/config/kube/crd/conversion.go [hunk 17]
```diff
@@ -78,10 +95,14 @@ func ConvertObject(schema collection.Schema, object IstioObject, domain string)
 	if err != nil {
 		return nil, err
 	}
-	data, err := FromJSON(schema, string(js))
+	spec, err := FromJSON(schema, string(js))
 	if err != nil {
 		return nil, err
 	}
+	status, err := IstioStatusJSONFromMap(object.GetStatus())
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/crd/conversion.go:102]: The status here is a `map`, and the Status in L120 is an interface, why do we even need to convert?

**Comment by hzxuzhonghu** [comment]
generally lgtm, need to address some nits

File: pilot/pkg/config/kube/crd/conversion_test.go (hunk 10 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/crd/conversion_test.go:52]: Can you update this?

File: pilot/pkg/config/kube/crdclient/client.go [hunk 18]
```diff
@@ -205,25 +205,36 @@ func (cl *Client) Get(typ config.GroupVersionKind, name, namespace string) *conf
...
-	meta, err := update(cl.istioClient, cl.serviceApisClient, config, getObjectMetadata(config))
+	meta, err := update(cl.istioClient, cl.serviceApisClient, cfg, getObjectMetadata(cfg))
+	if err != nil {
+		return "", err
+	}
+	return meta.GetResourceVersion(), nil
+}
+
+func (cl *Client) UpdateStatus(cfg config.Config) (string, error) {
+	if cfg.Spec == nil {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client.go:234]: shouldn't this not matter? we only care about status?

File: pilot/pkg/config/kube/crdclient/client_test.go [hunk 19]
```diff
@@ -125,7 +129,7 @@ func TestClient(t *testing.T) {
 
 			pb, err := r.NewInstance()
 			if err != nil {
-				t.Fatal(err)
+				t.Error(err)
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client_test.go:132]: nit: why the change? Fatal short circuits, I think the rest of the test is not valid if this fails

File: pilot/pkg/config/kube/crdclient/types.gen.go [hunk 20]
```diff
@@ -205,6 +206,103 @@ func update(ic versionedclient.Interface, sc serviceapisclient.Interface, cfg co
 	}
 }
 
+func updateStatus(ic versionedclient.Interface, sc serviceapisclient.Interface, cfg config.Config, objMeta metav1.ObjectMeta) (metav1.Object, error) {
+	switch cfg.GroupVersionKind {
+	case collections.IstioNetworkingV1Alpha3Destinationrules.Resource().GroupVersionKind():
+		return ic.NetworkingV1alpha3().DestinationRules(cfg.Namespace).UpdateStatus(context.TODO(), &clientnetworkingv1alpha3.DestinationRule{
+			ObjectMeta: objMeta,
+			Spec:       *(cfg.Spec.(*networkingv1alpha3.DestinationRule)),
+			Status:     *(cfg.Status.(*metav1alpha1.IstioStatus)),
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/types.gen.go:215]: I think status may be `nil` here since we don't check it in UpdateStatus. Spec doesn't have the same issue since we do check it explicitly there.

File: pkg/config/schema/codegen/collections.go [hunk 21]
```diff
@@ -67,8 +68,11 @@ var (
 			Plural: "{{ .Resource.Plural }}",
 			Version: "{{ .Resource.Version }}",
 			Proto: "{{ .Resource.Proto }}",
+			{{ if ne .Resource.StatusProto "" }}StatusProto: "{{ .Resource.StatusProto }}",{{end}}
```
Comment (howardjohn) [inline pkg/config/schema/codegen/collections.go:71]:
```suggestion
			{{- if ne .Resource.StatusProto "" }}StatusProto: "{{ .Resource.StatusProto }}",{{end}}
```
Not a big deal, but it means it will not insert a blank line if it is false

File: pkg/config/schema/metadata.yaml [hunk 22]
```diff
@@ -312,20 +316,26 @@ resources:
...
+    statusProto: "k8s.io.service_apis.api.v1alpha1.HTTPRouteStatus"
+    statusProtoPackage: "sigs.k8s.io/service-apis/apis/v1alpha1"
 
   - Kind: "TCPRoute"
     plural: "tcproutes"
     group: "networking.x-k8s.io"
     version: "v1alpha1"
     protoPackage: "sigs.k8s.io/service-apis/apis/v1alpha1"
     proto: "k8s.io.service_apis.api.v1alpha1.TCPRouteSpec"
+    statusProto: "k8s.io.service_apis.api.v1alpha1.TCPRouteStatus"
```
Comment (howardjohn) [inline pkg/config/schema/metadata.yaml:337]: Nice!!

**Review by howardjohn — COMMENTED** [review]
This is great, thanks!

@therealmitchconnors @jasonwzm can you take a look?

cc @stevenctl this will probably impact your work

File: releasenotes/notes/status-to-object-store.yaml [hunk 23]
```diff
@@ -0,0 +1,6 @@
+apiVersion: release-notes/v2
+kind: feature
+area: environments
+releaseNotes:
+  - |
+    **Added** support for status subresource in Istio Objects.
```
Comment (therealmitchconnors) [inline releasenotes/notes/status-to-object-store.yaml:6]: This note is not valid.  Status subresources have been supported since 1.6.

File: pilot/pkg/config/memory/config.go [hunk 24]
```diff
@@ -233,3 +233,7 @@ func (cr *store) Update(cfg config.Config) (string, error) {
 	ns.Store(cfg.Name, cfg)
 	return rev, nil
 }
+
+func (cr *store) UpdateStatus(cfg config.Config) (string, error) {
+	return "", nil
```
Comment (therealmitchconnors) [inline pilot/pkg/config/memory/config.go:238]: Is there a reason in memory resources can't have status?

File: pilot/pkg/serviceregistry/mcp/controller.go [hunk 25]
```diff
@@ -312,6 +312,11 @@ func (c *controller) Update(config.Config) (newRevision string, err error) {
 	return "", errUnsupported
 }
 
+func (c *controller) UpdateStatus(config.Config) (string, error) {
+	log.Warnf("updateStatus: %s", errUnsupported)
+	return "", errUnsupported
```
Comment (therealmitchconnors) [inline pilot/pkg/serviceregistry/mcp/controller.go:317]: Why is this error different from the "not implemented" error in the other classes?

File: pilot/pkg/model/fake_store.go [hunk 26]
```diff
@@ -71,6 +71,8 @@ func (s *FakeStore) Create(cfg config.Config) (revision string, err error) {
 
 func (*FakeStore) Update(config config.Config) (newRevision string, err error) { return "", nil }
 
+func (*FakeStore) UpdateStatus(config config.Config) (string, error) { return "", nil }
```
Comment (therealmitchconnors) [inline pilot/pkg/model/fake_store.go:74]: I can't find anywhere this function is actually called, outside of its own implementation and tests.  Where is this function going to be used?  Is its usage going to be in a separate PR?

**Review by therealmitchconnors — CHANGES_REQUESTED** [review]
I've taken a first pass at review, but it's very hard to understand the purpose of these changes without context.  Can you please link to your design doc and tracking issue, and I'll make a more detailed review pass.

File: pilot/pkg/model/fake_store.go (hunk 26 above)
Comment (adiprerepa) [reply pilot/pkg/model/fake_store.go:74]: yes, in VM healthchecking

Comment (howardjohn) [reply pilot/pkg/model/fake_store.go:74]: Yes, we have two designs that will use it: auto registration and vm health checking. https://docs.google.com/document/d/10TyPy-6bOs0_7JljGpc2RneDdsaR9GktjwwaaVbz6fs/edit for more info

**Comment by adiprerepa** [comment]
@therealmitchconnors this pr doesnt have a design doc, but it was made so [vm healthchecking ](https://docs.google.com/document/d/1G4s_PrmvkwCKoH_kXVeSUAszsaj3ZfpIlraSdeTSPKU/) could update `WorkloadEntry` statuses. The istio-agent part has been implemented, but workloadentry status needs to be updated on health change. we say we do this in the `WorkloadEntry` [docs](https://istio.io/latest/docs/reference/config/networking/workload-entry/), but we really dont.

File: releasenotes/notes/status-to-object-store.yaml (hunk 23 above)
Comment (adiprerepa) [reply releasenotes/notes/status-to-object-store.yaml:6]: so no release notes?

File: pilot/pkg/config/kube/ingress/controller.go [hunk 27]
```diff
@@ -318,6 +318,10 @@ func (c *controller) Update(_ config.Config) (string, error) {
 	return "", errUnsupportedOp
 }
 
+func (c *controller) UpdateStatus(config.Config) (string, error) {
+	return "", nil
```
Comment (stevenctl) [inline pilot/pkg/config/kube/ingress/controller.go:322]: Should this return `errUnsupportedOp`?

File: pilot/pkg/config/memory/config.go (hunk 24 above)
Comment (stevenctl) [reply pilot/pkg/config/memory/config.go:238]: This would probably be useful in tests.

File: pilot/pkg/serviceregistry/mcp/controller.go (hunk 25 above)
Comment (adiprerepa) [reply pilot/pkg/serviceregistry/mcp/controller.go:317]: not sure. just mirrored what the other controller methods did.

File: pilot/pkg/config/kube/ingress/controller.go (hunk 27 above)
Comment (adiprerepa) [reply pilot/pkg/config/kube/ingress/controller.go:322]: indeed it should be

File: releasenotes/notes/status-to-object-store.yaml (hunk 23 above)
Comment (stevenctl) [reply releasenotes/notes/status-to-object-store.yaml:6]: I'm not sure if this PR needs a note - it's sort of an internal change to enable other work. 

As far as supporting status resources since 1.6... how are we writing to status currently?

File: pilot/pkg/serviceregistry/mcp/controller.go (hunk 25 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/mcp/controller.go:317]: It's a little different. We have configs stores that are "writers" which are only memory and k8s. Everything else is read-only by design IIUC.

File: pilot/pkg/config/kube/crdclient/gen/types.go.tmpl [hunk 28]
```diff
@@ -69,6 +70,22 @@ func update(ic versionedclient.Interface, sc serviceapisclient.Interface, cfg co
 	}
 }
 
+func updateStatus(ic versionedclient.Interface, sc serviceapisclient.Interface, cfg config.Config, objMeta metav1.ObjectMeta) (metav1.Object, error) {
+    switch cfg.GroupVersionKind {
+    {{- range . }}
+    	case collections.{{ .VariableName }}.Resource().GroupVersionKind():
+    		return {{.Client}}.{{ .ClientGroupPath }}().{{ .ClientTypePath }}({{if .Namespaced}}cfg.Namespace{{end}}).UpdateStatus(context.TODO(), &{{ .ClientImport }}.{{ .Kind }}{
+    			ObjectMeta: objMeta,
+    			Spec:       *(cfg.Spec.(*{{ .APIImport }}.{{ .Kind }}{{ .TypeSuffix }})),
```
Comment (stevenctl) [inline pilot/pkg/config/kube/crdclient/gen/types.go.tmpl:79]: Does the `UpdateStatus` method here for client-go modify the spec _and_ status?

Comment (stevenctl) [reply pilot/pkg/config/kube/crdclient/gen/types.go.tmpl:79]: https://github.com/istio/client-go/blob/master/pkg/clientset/versioned/typed/networking/v1alpha3/workloadentry.gen.go#L139

Looks like this would only affect the subresource.

**Review by stevenctl — APPROVED** [review]

File: releasenotes/notes/status-to-object-store.yaml (hunk 23 above)
Comment (therealmitchconnors) [reply releasenotes/notes/status-to-object-store.yaml]: So this PR doesn't write to the status field, it just lays the groundwork for another PR to write to the status field?

Here is the existing mechanism for writing to the status field.  We need to ensure this work will not overwrite this existing status.  https://istio.io/latest/docs/reference/config/config-status/

File: tests/integration/pilot/analysis/analysis_test.go [hunk 29]
```diff
@@ -87,12 +91,81 @@ spec:
...
+					},
+				},
+			}
+			ctx.Clusters().Default().Istio().NetworkingV1alpha3().WorkloadEntries(ns.Name()).UpdateStatus(context.TODO(), &clientnetworkingv1alpha3.WorkloadEntry{
+				Status:  *stat,
+
+			}, metav1.UpdateOptions{})
+			retry.UntilSuccessOrFail(t, func() error {
+				// should update
+				return expectWorkloadEntryStatus(t, ctx, ns, stat.Conditions)
```
Comment (therealmitchconnors) [inline tests/integration/pilot/analysis/analysis_test.go:142]: The purpose of this test is to check compatibility with the CRD Status work from 1.6.  As such, the flow should look like: 

1. Create Workload Entry
2. Check for condition reconciled=true
3. Update Status to includ conditino Healthy=true
4. Check that condition reconciled=true and healthy=true

This will prove that this mechanism won't interfere with existing conditions.

Comment (adiprerepa) [reply tests/integration/pilot/analysis/analysis_test.go:142]: where does reconciled come from ehre?

Comment (howardjohn) [reply tests/integration/pilot/analysis/analysis_test.go:142]: https://github.com/istio/istio/blob/b25c7eb039dc422cb47fc6015f38bf9236dd5208/pilot/pkg/status/state.go#L243

**Comment by adiprerepa** [comment]
ignore the 800 million commits that a rebase accidentally summoned

**Comment by adiprerepa** [comment]
@therealmitchconnors the integration test "works" now, but its not getting a status reconciled. Did I test it the right way?

File: tests/integration/pilot/analysis/analysis_test.go (hunk 29 above)
Comment (adiprerepa) [reply tests/integration/pilot/analysis/analysis_test.go:157]: It looks like reconciled is not created when I directly call Create(). Do I need to create it some other way for status to be Reconciled?

File: tests/integration/pilot/analysis/analysis_test.go [hunk 30]
```diff
@@ -106,41 +105,77 @@ func TestWorkloadEntryUpdatesStatus(t *testing.T) {
...
-			retry.UntilSuccessOrFail(t, func() error {
-				// should update
-				return expectWorkloadEntryStatus(t, ctx, ns, stat.Conditions)
-			}, retry.Timeout(time.Minute*5))
+			// get the workload entry to replace the health condition field
+			we, err = ctx.Clusters().Default().Istio().NetworkingV1alpha3().WorkloadEntries(ns.Name()).Get(context.TODO(), "vm-1", metav1.GetOptions{})
 
 			// replacing the condition
-			for i, cond := range stat.Conditions {
+			for i, cond := range we.Status.Conditions {
```
Comment (therealmitchconnors) [inline tests/integration/pilot/analysis/analysis_test.go:176]: This logic should be abstracted into a public function, to ensure this pattern is used for updating conditions in the feature.

File: tests/integration/pilot/analysis/analysis_test.go [hunk 31]
```diff
@@ -106,41 +105,77 @@ func TestWorkloadEntryUpdatesStatus(t *testing.T) {
...
+spec:
+  address: 127.0.0.1
 `)
-			// Test adding one condition
-			stat := &v1alpha1.IstioStatus{
-				Conditions: []*v1alpha1.IstioCondition{
+
+			retry.UntilSuccessOrFail(t, func() error {
+				// we should expect an empty array not nil
+				return expectWorkloadEntryStatus(t, ctx, ns, nil)
```
Comment (therealmitchconnors) [inline tests/integration/pilot/analysis/analysis_test.go:121]: We should expect that condition Reconciled and condition PassedAnalysis per https://istio.io/latest/docs/reference/config/config-status/.  Is that not what you are seeing?

Comment (adiprerepa) [reply tests/integration/pilot/analysis/analysis_test.go:121]: no, im seeing a nil condition array for some reason. Is this an issue with my test?

Do I have to call something to trigger those conditions?

File: tests/integration/pilot/analysis/analysis_test.go (hunk 30 above)
Comment (adiprerepa) [reply tests/integration/pilot/analysis/analysis_test.go:176]: I think adding that function once healthchecking is being implemented and replacing this logic with that function might be the better move. I might be wrong.

**Review by therealmitchconnors — APPROVED** [review]

File: pilot/pkg/config/kube/crdclient/gen/main.go [hunk 32]
```diff
@@ -128,7 +133,7 @@ func main() {
 	// Format source code.
 	out, err := format.Source(buffer.Bytes())
 	if err != nil {
-		log.Fatal(err)
+		log.Print(buffer.String())
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/gen/main.go:136]: can you revert this change?

File: pilot/pkg/config/kube/gateway/conversion_test.go (hunk 12 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion_test.go:57]: Please remove

File: pilot/pkg/config/kube/gateway/testdata/http.yaml_BACKUP_222124.golden [hunk 33]
```diff
@@ -0,0 +1,94 @@
+apiVersion: networking.istio.io/v1alpha3
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/testdata/http.yaml_BACKUP_222124.golden:1]: can you remove this?

File: pilot/pkg/config/kube/gateway/testdata/http.yaml_BASE_222124.golden [hunk 34]
```diff
@@ -0,0 +1,87 @@
+apiVersion: networking.istio.io/v1alpha3
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/testdata/http.yaml_BASE_222124.golden:1]: can you remove?

File: pilot/pkg/config/kube/gateway/testdata/http.yaml_LOCAL_222124.golden [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+  namespace: istio-system
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+  - hosts:
+    - '*.domain.example'
+    port:
+      name: http-80-gateway-gateway-istio-system
+      number: 80
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/testdata/http.yaml_LOCAL_222124.golden:15]: can you remove?

**Review by howardjohn — COMMENTED** [review]
Looks good!! can you just clean up the extra files and we can merge this. Thanks @adiprerepa !

File: pilot/pkg/config/kube/gateway/testdata/http.yaml_REMOTE_222124.golden [hunk 36]
```diff
@@ -0,0 +1,91 @@
+apiVersion: networking.istio.io/v1alpha3
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/testdata/http.yaml_REMOTE_222124.golden:1]: please remove

**Review by howardjohn — APPROVED** [review]
LGTM
